volatility3/framework/layers/qemu.py

```diff
@@ -74,32 +111,43 @@ def _get_ram_segments(self, index: int, page_size: int) -> Tuple[List[Tuple[int,
         done = None
         segments = []
 
+        size_array = {}
         base_layer = self.context.layers[self._base_layer]
 
         while not done:
-            addr = self.context.object(self._qemu_table_name + constants.BANG + 'unsigned long long',
-                                       offset = index,
-                                       layer_name = self._base_layer)
+            # Use struct.unpack here for performance improvements
+            addr = struct.unpack('>Q', base_layer.read(index, 8))[0]
+
             # Flags are stored in the n least significant bits, where n equals the bit-length of pagesize
             flags = addr & (page_size - 1)
             # addr equals the highest multiple of pagesize <= offset
             # (We assume that page_size is a power of 2)
             addr = addr ^ (addr & (page_size - 1))
             index += 8
 
+            if addr >= self._pci_hole_start:
+                addr += self._pci_hole_end - self._pci_hole_start
+
             if flags & self.SEGMENT_FLAG_MEM_SIZE:
                 namelen = self._context.object(self._qemu_table_name + constants.BANG + 'unsigned char',
                                                offset = index,
                                                layer_name = self._base_layer)
                 while namelen != 0:
-                    # if base_layer.read(index + 1, namelen) == b'pc.ram':
-                    #     total_size = self._context.object(self._qemu_table_name + constants.BANG + 'unsigned long long',
-                    #                                       offset = index + 1 + namelen,
-                    #                                       layer_name = self._base_layer)
+                    total_size = self._context.object(self._qemu_table_name + constants.BANG + 'unsigned long long',
+                                                      offset = index + 1 + namelen,
+                                                      layer_name = self._base_layer)
+                    size_array[base_layer.read(index + 1, namelen)] = total_size
                     index += 1 + namelen + 8
                     namelen = self._context.object(self._qemu_table_name + constants.BANG + 'unsigned char',
                                                    offset = index,
                                                    layer_name = self._base_layer)
+                if size_array.get(b'pc.ram',
```

Comment:
Perhaps replace this with `if b'pc.ram' in size_array and size_array[b'pc.ram'] < self._pci_hole_minimum:` for better readability? Moreover, it needs to be `<` instead of `<=` I think?

Reply:
Using defaults is fairly standard throughout our codebase, but I've moved the complex bit into a variable for clarity.